node/ipc.go

```diff
@@ -0,0 +1,375 @@
+package node
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"log"
+	"math/rand"
+	"net"
+	"net/url"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/INFURA/ethereum-interaction/pkg/jsonrpc"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+)
+
+func newIPCTransport(ctx context.Context, parsedURL *url.URL) (transport, error) {
+	conn, err := net.Dial("unix", parsedURL.String())
+	if err != nil {
+		return nil, errors.Wrap(err, "could not connect over IPC")
+	}
+
+	t := ipcTransport{
+		conn:                   conn,
+		ctx:                    ctx,
+		counter:                rand.Uint64(),
+		chToBackend:            make(chan jsonrpc.Request),
+		chSubscriptionRequests: make(chan *subscriptionRequest),
+		chOutboundRequests:     make(chan *outboundRequest),
+		subscriptonRequests:    make(map[jsonrpc.ID]*subscriptionRequest),
+		outboundRequests:       make(map[jsonrpc.ID]*outboundRequest),
+		subscriptions:          make(map[string]*subscription),
+	}
+
+	go t.loop()
+	return &t, nil
+}
+
+type ipcTransport struct {
+	conn net.Conn
+	ctx  context.Context
+
+	counter                uint64
+	chToBackend            chan jsonrpc.Request
+	chSubscriptionRequests chan *subscriptionRequest
+	chOutboundRequests     chan *outboundRequest
+
+	subscriptonRequests map[jsonrpc.ID]*subscriptionRequest
+	outboundRequests    map[jsonrpc.ID]*outboundRequest
+	requestMu           sync.RWMutex
+
+	subscriptions   map[string]*subscription
+	subscriptionsMu sync.RWMutex
+}
+
+func (t *ipcTransport) loop() {
+	g, ctx := errgroup.WithContext(t.ctx)
```

Comment:
This function is currently more or less a copy/paste from the websocket code, should be refactored into common functionality.  I think a generic "loop" that took an interface w/ these methods:

```golang
type nameTBD interface {
  ReadMessage() ([]byte, error)
  WriteMessage([]byte) error
}
```

Would allow the code to be shared.  For example the IPC implementation would be something like:

```golang
  ReadMessage := func() { 
     if !scanner.Scan() { return nil, scanner.Err() }
     return []byte(scanner.Text()), nil
  }
  WriteMessage :=  func() {
     conn.Write(...)
  }
```

Reply:
63e7220 pulls the differences into two closures, now I just need to decide where `loop` should live and how the closures get injected.

Reply:
Got a little time to dust this off, refactored the two into a shared loopingTransport that handles most of the heavy lifting.